### ISSUE TITLE
Upgrade from 1.4.10 to 2.0.0 does not work

### ISSUE THREAD

**Issue by sirfranz** [evolution-cms/evolution#1360]
I can't upgrade evo from 1.4.10 to 2.0.0.

I get the error: Warning: count(): Parameter must be an array or an object that implements Countable in .../html/install/src/functions.php on line 460

If I retry the setup, it shows that it was sucessfull. But evo does'nt work anymore. It shows on every page "Error" even at the admin section.

**Comment by BBloke** [comment]
It sounds like there is a snippet running in your template/page which may be throwing the error.  I've had this.

What I usually do is rename all the snippets in my template, save and see if the site works and then rename them one by one until it breaks to find the culprit.

The preference in running v2.0 is to use DocLister over Ditto, FormLister over WebLoginPE, evoSearch over AjaxSearch.

I successfully managed to upgrade from 1.4.10 to 2.0 with a fair bit of ground work to begin with.  It can be done and to be honest quite easy.  The instructions and video I found were good enough to get me there.

**Comment by sirfranz** [comment]
@BBloke, thanks for your answer, but unfortunately it does not help me. The error also occurs in the admin area. So I even can not login as admin. So I can not change anything at the admin section. Because only "Error" is displayed, I have no clue what the problem is.

**Comment by BBloke** [comment]
Do a check list.

PHP version. Composer. Do you have an Evo dB file for the dB settings?

**Comment by sirfranz** [comment]
PHP version:  	7.2.26
Composer:        Is it not automatically installed by the installer?
DB file:              core/config/database/connections/default.php, exists (I have replaced sensitive data in den Listing below by xxx):

```
<?php
return [
    'driver' => env('DB_TYPE', 'mysqli'), //$database_type
    'host' => env('DB_HOST', 'localhost'), //$database_server
    'port' => env('DB_PORT', '3306'),
    'database' => env('DB_DATABASE', 'xxx'), //$dbase
    'username' => env('DB_USERNAME', 'xxx'), //$database_user
    'password' => env('DB_PASSWORD', 'xxx'), //$database_password
    'unix_socket' => env('DB_SOCKET', ''),
    'charset' => env('DB_CHARSET', 'utf8'), // $database_connection_charset
    'collation' => env('DB_COLLATION', 'latin1_german1_ci'), //$database_collation
    'prefix' => env('DB_PREFIX', 'modx_'), //$table_prefix
    'method' => env('DB_METHOD', 'SET NAMES'), //$database_connection_method
    'strict' => env('DB_STRICT', false),
    'engine' => env('DB_ENGINE'),
    'options' => [
        PDO::ATTR_STRINGIFY_FETCHES => true,
    ]
];
```

**Comment by fourroses666** [comment]
"Error" probably means it can't find the database. You need to edit your config file I think which probably is located on a different location.

**Comment by BBloke** [comment]
You don't necessarily need composer to install.  Although I have it on a windows machine running apache 2.4 and it's ok.  Whilst I've tried to install it on a NAS box I never get anything showing on manager or site.

Did you create the .install file in the /core/ folder?

I remember I found it easier to make sure I got the right DB settings in the new file by creating a new installation so it does all the work for me.

Out of interest try changing the DB from mysqli to mysql.  I think I had trouble with mysqli.

**Comment by sirfranz** [comment]
Is there a possibility that the full PHP error message is displayed instead of "Error" even without login?

**Comment by sirfranz** [comment]
@BBloke, I have created the .install file in the /core/ folder.

**Comment by BBloke** [comment]
Normally when I 've seen an "Error" on a page the answer can sometimes lie in the PHP Error file.

I've just tried to use the installer to whip up a new site on my sysnology box running apache 2.4 and php 7.3 bu it doesn't finish the installer although everything looks like it is in place.  I think v2.0 doesn't like the webs erver config on sysnology.

I'm running out of ideas.  Have you tried the DB change from mysqli to mysql?

**Comment by sirfranz** [comment]
Changing mysqli to mysql has not solved the problem.

**Comment by BBloke** [comment]
Have you tried removing all the snippet calls from your template?  I usually just put an underscore at the beginning so it doesn't run.

**Comment by sirfranz** [comment]
Many thanks to all.
Now it's nearly working.

I have modified the ExceptionHandler.php, so I could read what the problem is: 
````« Evolution CMS Parse Error »
Cannot declare class PHxParser, because the name is already in use
Error information	
File	/var/www/xxx/html/assets/plugins/phx/phx.parser.class.inc.php
Line	12
```` 

I have modified the phx.parser.class.inc.php:
````if (!class_exists('PHxParser')){
class PHxParser {
	var $placeholders = array();
	
	//old: function PHxParser($debug=0,$maxpass=50) {
	function __construct($debug=0,$maxpass=50) {
````

and the site works at the front end.

But now, my next problem is: The manager dashboard is not working. It shows:
````« Evolution CMS Parse Error »
Using $this when not in object context (View: /var/www/xxx/html/manager/views/page/2.blade.php)
Error information 	
File 	/var/www/xxx/html/assets/cache/siteCache.idx.php
Line 	2````

**Comment by sirfranz** [comment]
Now I have deleted the siteCache.idx.php
and the manager dashboard works too.

But there is a new problem:
All pages with umlauts (ä,ö,ü,ß) in title has chartset errors. The menu description ist not showing. And this pages are no longer editable:

````
« Evolution CMS Parse Error »

Error : mb_convert_encoding(): Unable to detect character encoding

Error information 	
ErrorType[num] 	WARNING[2]
File 	/var/www/xxx/html/core/src/Legacy/PhpCompat.php
Line 	38
Source 	$ent_str = mb_convert_encoding($str, $encode, $detect_order); 
````

**Comment by BBloke** [comment]
Glad to read you are having some success. Sadly, I assumed you'd have found the instructions which advises you delete sitecache. Sorry! 

Do you use PHx?

I was advised to remove PHX from my site and use IF instead. I didn't find it too difficult to switch over especially as there was a custom webGroup handler for IF. I'd recommend investing some time if you do. 

I noticed in the system events lots of PHX 2.2.0 event logs until I disabled the plugin.

I'm afraid I cannot help on the special characters as everything I've worked on is in plain English.

Good luck. I hope everything works as needed quickly.
